File: analysis/discursos_presidenciales/03_kast_alusiones.R
#!/usr/bin/env Rscript
# Discursos Kast: léxico, alusiones, Kast vs Kaiser — figuras ggplot2
#
# Lee los CSV de 03_kast_alusiones.py (cuerpo oral + prensa unida).
# Si faltan, llama al .py y sigue.
#
# Uso:
#   python3 analysis/discursos_presidenciales/03_kast_alusiones.py
#   Rscript analysis/discursos_presidenciales/03_kast_alusiones.R

suppressPackageStartupMessages({
  library(tidyverse)
  library(scales)
  library(lubridate)
})

for (p in c("analysis/_helpers.R", "../_helpers.R")) {
  if (file.exists(p)) { source(p, local = FALSE); break }
}

root <- project_root()
out_fig <- out_imagenes(root)
canon <- canon_dir(root)
prensa_dir <- file.path(root, "data", "processed", "prensa")

need <- c(
  file.path(canon, "discursos_kast_alusiones.csv"),
  file.path(canon, "discursos_kast_top_palabras.csv"),
  file.path(canon, "discursos_kast_repertorios.csv"),
  file.path(canon, "discursos_kast_kaiser_mes.csv"),
  file.path(prensa_dir, "prensa_kast_kaiser_anio.csv")
)
if (any(!file.exists(need))) {
  message("Faltan CSV — corriendo 03_kast_alusiones.py …")
  st <- system2("python3", file.path(root, "analysis/discursos_presidenciales/03_kast_alusiones.py"))
  if (!identical(st, 0L)) stop("Falló el .py (hace falta para los CSV)")
}

theme_ng <- theme_minimal(base_size = 12) +
  theme(
    plot.title = element_text(face = "bold", size = 13),
    plot.subtitle = element_text(color = "grey40", size = 10),
    panel.grid.minor = element_blank(),
    legend.position = "bottom"
  )

col_kast <- "#1B4F72"
col_kaiser <- "#D69E2E"
col_alus <- "#C53030"
fam_col <- c(A = "#E8A598", C = "#3D9B64", D = "#4C78A8")

al <- read_csv(file.path(canon, "discursos_kast_alusiones.csv"), show_col_types = FALSE)
top <- read_csv(file.path(canon, "discursos_kast_top_palabras.csv"), show_col_types = FALSE)
rep <- read_csv(file.path(canon, "discursos_kast_repertorios.csv"), show_col_types = FALSE)
mes <- read_csv(file.path(canon, "discursos_kast_kaiser_mes.csv"), show_col_types = FALSE) |>
  mutate(mes = as.Date(mes))
anio <- read_csv(file.path(prensa_dir, "prensa_kast_kaiser_anio.csv"), show_col_types = FALSE)
n_disc <- unique(al$n_discursos)[[1]]

ruido <- c("haber", "habia", "quien", "otro", "otros", "solamente", "haciendo", "adelante")
p_top <- top |>
  filter(!palabra %in% ruido) |>
  slice_head(n = 22) |>
  mutate(palabra = fct_reorder(palabra, n)) |>
  ggplot(aes(n, palabra)) +
  geom_col(fill = col_kast, width = 0.75) +
  labs(
    title = "Qué palabras nombra Kast (cuerpo de los discursos)",
    subtitle = sprintf("n = %d discursos · 11 mar – 30 jul 2026 · titular de Presidencia excluido", n_disc),
    x = "Frecuencia", y = NULL
  ) +
  theme_ng

ggsave(file.path(out_fig, "discursos_kast_top_palabras.png"), p_top,
       width = 9, height = 7.2, dpi = 140)

p_al <- al |>
  mutate(etiqueta = fct_reorder(etiqueta, pct)) |>
  ggplot(aes(pct, etiqueta)) +
  geom_col(fill = col_alus, width = 0.72) +
  geom_text(aes(label = sprintf("%.0f%%  (%d)", pct, n)), hjust = -0.05, size = 3) +
  coord_cartesian(xlim = c(0, max(al$pct) * 1.22)) +
  labs(
    title = "A qué hace alusión Kast",
    subtitle = "Porcentaje de discursos cuyo cuerpo menciona cada tema o nombre",
    x = "% de discursos", y = NULL
  ) +
  theme_ng

ggsave(file.path(out_fig, "discursos_kast_alusiones.png"), p_al,
       width = 9.5, height = 7.4, dpi = 140)

mes_l <- mes |>
  select(mes, pct_kast, pct_kaiser) |>
  pivot_longer(-mes, names_to = "quien", values_to = "pct") |>
  mutate(quien = recode(quien,
    pct_kast = "dice “Kast”",
    pct_kaiser = "dice “Kaiser”"
  ))

p_mes <- ggplot(mes_l, aes(mes, pct, color = quien)) +
  geom_line(linewidth = 1.05) +
  geom_point(size = 2.6) +
  scale_color_manual(values = c("dice “Kast”" = col_kast, "dice “Kaiser”" = col_kaiser)) +
  scale_x_date(date_breaks = "1 month", date_labels = "%b") +
  labs(
    title = "¿Kast se nombra? ¿Nombra a Kaiser? (cuerpo oral, por mes)",
    subtitle = "El titular de sala de prensa no cuenta. Kaiser aparece una sola vez (Enela, jun-2026).",
    x = NULL, y = "% de discursos del mes", color = NULL
  ) +
  theme_ng

ggsave(file.path(out_fig, "discursos_kast_kaiser_mes.png"), p_mes,
       width = 10, height = 5.2, dpi = 140)

p_rep <- rep |>
  mutate(lab = paste(codigo, etiqueta), lab = fct_reorder(lab, pct)) |>
  ggplot(aes(lab, pct, fill = familia)) +
  geom_col(width = 0.72) +
  coord_flip() +
  scale_fill_manual(values = fam_col) +
  labs(
    title = "Repertorios A–D en la boca de Kast",
    subtitle = sprintf("n = %d  ·  cuerpo oral", n_disc),
    x = NULL, y = "% discursos", fill = "Familia"
  ) +
  theme_ng

ggsave(file.path(out_fig, "discursos_kast_repertorios.png"), p_rep,
       width = 10, height = 5.8, dpi = 140)

anio_pct <- anio |>
  select(year, pct_kast, pct_kaiser) |>
  pivot_longer(-year, names_to = "quien", values_to = "pct") |>
  mutate(quien = recode(quien,
    pct_kast = "titular con Kast",
    pct_kaiser = "titular con Kaiser"
  ))

p_pk <- ggplot(anio_pct, aes(year, pct, color = quien)) +
  geom_line(linewidth = 1.05) +
  geom_point(size = 2.5) +
  scale_x_continuous(breaks = 2015:2026) +
  scale_color_manual(values = c(
    "titular con Kast" = col_kast,
    "titular con Kaiser" = col_kaiser
  )) +
  labs(
    title = "Prensa (2015–2026): ¿se nombra a Kast o a Kaiser?",
    subtitle = "Prensa 2015–2026. Porcentaje sobre el total de notas del año.",
    x = NULL, y = "% de titulares", color = NULL
  ) +
  theme_ng +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(file.path(out_fig, "prensa_kast_kaiser_anio.png"), p_pk,
       width = 11, height = 5.4, dpi = 140)

anio_n <- anio |>
  select(year, n_kast, n_kaiser) |>
  pivot_longer(-year, names_to = "quien", values_to = "n") |>
  mutate(quien = recode(quien, n_kast = "Kast", n_kaiser = "Kaiser"))

p_pn <- ggplot(anio_n, aes(year, n, fill = quien)) +
  geom_col(position = "dodge", width = 0.75) +
  scale_x_continuous(breaks = 2015:2026) +
  scale_y_continuous(labels = comma) +
  scale_fill_manual(values = c(Kast = col_kast, Kaiser = col_kaiser)) +
  labs(
    title = "Volumen de titulares: Kast vs Kaiser",
    subtitle = "Misma prensa. 2025–26: Kast dispara; Kaiser no lo acompaña al mismo ritmo.",
    x = NULL, y = "Titulares", fill = NULL
  ) +
  theme_ng +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(file.path(out_fig, "prensa_kast_kaiser_volumen.png"), p_pn,
       width = 11, height = 5.4, dpi = 140)

p_vol <- ggplot(anio, aes(year, n)) +
  geom_col(fill = col_kast, width = 0.72) +
  scale_x_continuous(breaks = 2015:2026) +
  scale_y_continuous(labels = comma) +
  labs(
    title = "Prensa, 2015–2026",
    subtitle = "Notas por año. Una sola serie.",
    x = NULL, y = "Notas"
  ) +
  theme_ng +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(file.path(out_fig, "prensa_unida_anio.png"), p_vol,
       width = 11, height = 5.2, dpi = 140)

message("ggplot2 → ", out_fig)
print(al |> arrange(desc(pct)), n = 30)
print(anio)

File: analysis/discursos_presidenciales/03_kast_alusiones.py
#!/usr/bin/env python3
"""Discursos presidenciales de Kast: léxico, alusiones y tendencia Kaiser vs Kast.

Usa el *cuerpo* oral (no el titular de Presidencia, que casi siempre dice “Kast”).
Gobierno Kast = 2026-03-11 en adelante. Prensa = corpus unido 2015–2026 (títulos).

CSV para ggplot2 (`03_kast_alusiones.R`). Uso (desde la raíz):
  python3 analysis/discursos_presidenciales/03_kast_alusiones.py
  Rscript analysis/discursos_presidenciales/03_kast_alusiones.R
"""
from __future__ import annotations

import re
import sys
import unicodedata
from datetime import date
from pathlib import Path

import pandas as pd

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "data" / "scripts"))
from build_canonicos import GOBIERNO_KAST, fold, parse_fecha_es  # noqa: E402

OUT_CANON = ROOT / "data" / "processed" / "canon"
OUT_PRENSA = ROOT / "data" / "processed" / "prensa"
RAW_DISC = ROOT / "data" / "raw" / "discursos" / "presidencia" / "discursos.csv"
PRENSA_UNIDA = ROOT / "data" / "processed" / "prensa" / "prensa_unida.parquet"

STOP = {
    "de", "la", "el", "en", "y", "a", "los", "del", "las", "se", "un", "una",
    "que", "por", "con", "para", "como", "al", "lo", "su", "sus", "es", "mas",
    "no", "o", "tambien", "este", "esta", "estos", "estas", "hay", "son", "ser",
    "ha", "han", "hemos", "muy", "ya", "si", "pero", "cuando", "donde", "sobre",
    "entre", "desde", "hasta", "sin", "nos", "hoy", "aqui", "asi", "todo",
    "todos", "todas", "cada", "fue", "era", "tiene", "tienen", "hacer", "puede",
    "pueden", "debe", "estan", "estamos", "esto", "porque", "pues", "entonces",
    "ademas", "solo", "mismo", "chile", "chilenos", "chilenas", "pais",
    "gobierno", "presidente", "republica", "antonio", "jose", "boric",
    "gabriel", "tenemos", "nosotros", "ustedes", "usted", "anos", "ano", "dia",
    "dias", "vez", "veces", "menos", "mejor", "mayor", "sido", "siendo", "hace",
    "mucho", "estar", "bien", "tiempo", "poder", "punto", "caso", "parte",
    "manera", "forma", "hecho", "decir", "ahora", "despues", "antes", "siempre",
    "gran", "gracias", "vamos", "voy", "tema", "temas", "bueno", "buena",
    "muchas", "muchos", "quiero", "creo", "claro", "senor", "senora", "ministros",
    "ministro", "companeros", "amigo", "amigos", "queridos", "estimados",
    "algo", "alguien", "algunos", "algunas", "cosas", "cosa", "decia", "dice",
    "esos", "esas", "ese", "esa", "tanto", "distintas", "distintos", "nuestra",
    "nuestro", "nuestros", "nuestras", "tener", "podemos", "importante",
    "aqui", "alli", "ahi", "hacia", "traves", "mediante", "durante",
    "queremos", "querer", "queria", "querian",
    "necesitamos", "necesito", "necesita", "necesitan", "necesario",
    "haciendo", "haber", "habia", "habian", "habra",
    "quien", "quienes", "otro", "otros", "otra", "otras",
    "adelante", "solamente", "tambien",
    "nuevo", "nueva", "nuevos", "nuevas",
    "vida", "hacerlo", "podamos", "debemos", "tenemos",
}

# Prefijos: “necesita*”, “querem*”, etc. (ruido oral, no léxico de política)
STOP_PREFIX = (
    "necesit", "querem", "haciend", "haber", "habia", "adelant",
    "solament", "quien", "otro", "otra", "nuevo", "nueva",
)

# Cuerpo oral — no el titular de sala de prensa.
ALUSIONES = [
    ("empleo / trabajo", r"\bempleo\b|\btrabajo\b|\btrabajadores?\b"),
    ("seguridad", r"seguridad|delincuencia|narcotrafico|crimen organizado"),
    ("familia", r"\bfamilias?\b"),
    ("inversión", r"inversion|emprendedor|emprendimiento"),
    ("Carabineros", r"carabineros"),
    ("Fuerzas Armadas", r"fuerzas armadas|ejercito"),
    ("libertad", r"\blibertades?\b|libre mercado|libertad economica"),
    ("migración", r"migracion|migrantes|frontera"),
    ("orden", r"\borden\b|estado de derecho"),
    ("Dios / fe", r"\bdios\b|fe cristiana|\biglesia\b"),
    ("impuestos", r"impuesto"),
    ("Araucanía", r"araucania|macrozona"),
    ("sociedad civil", r"sociedad civil|cuerpos intermedios|juntas? de vecinos"),
    ("Venezuela", r"venezuela|\bmaduro\b"),
    ("gremios", r"gremialismo|gremialista|\bgremios\b"),
    ("Kast (se nombra)", r"\bkast\b"),
    ("Piñera", r"\bpinera\b"),
    ("Boric", r"\bboric\b"),
    ("permisología", r"permisolog"),
    ("Matthei", r"\bmatthei\b"),
    ("Kaiser", r"\bkaiser\b"),
    ("subsidiariedad", r"subsidiariedad"),
    ("Pinochet", r"\bpinochet\b"),
    ("Jaime Guzmán", r"jaime guzman"),
]

DICC = [
    ("A1", "A", "Subsidiariedad / cuerpos intermedios",
     r"subsidiariedad|cuerpos intermedios|gremios|bien comun|juntas? de vecinos|"
     r"gremialismo|gremialista|sociedad civil|tejido social|"
     r"organizaciones intermedias|asociatividad|principio de subsidiariedad"),
    ("A2", "A", "Orden institucional",
     r"orden publico|seguridad juridica|estado de derecho|estabilidad|autoridad|"
     r"mano dura|orden institucional|certeza juridica"),
    ("A3", "A", "Familia / moral",
     r"familia|matrimonio|valores|tradicion cristiana|dignidad humana|"
     r"familia tradicional|valores familiares"),
    ("A4", "A", "Antimarxismo / antiestatismo",
     r"asistencialismo|dependencia estatal|populismo|marxismo|estatismo|estado interventor"),
    ("A5", "A", "Economía libre",
     r"libre mercado|propiedad privada|emprendimiento|inversion|"
     r"subsidiariedad economica|libertad economica|economia libre"),
    ("C1", "C", "Securitización",
     r"narcotrafico|mano dura|tolerancia cero|delincuencia|crisis de seguridad|"
     r"seguridad|crimen organizado|terrorismo"),
    ("C2", "C", "Migración",
     r"migracion ilegal|crisis migratoria|frontera|migrantes|migracion|"
     r"inmigracion|extranjeros ilegales|control fronterizo"),
    ("C3", "C", "Guerra cultural",
     r"ideologia de genero|woke|adoctrinamiento|identidad de genero|"
     r"enfoque de genero|cultura woke|cancelacion|educacion sexual integral"),
    ("C4", "C", "Anti-élite",
     r"elite globalista|academia desconectada|casta politica|\bcasta\b|"
     r"globalismo|las elites|elite politica|pueblo versus elite"),
    ("C5", "C", "Civilización",
     r"valores occidentales|chile primero|occidente cristiano|hispanidad|"
     r"identidad nacional|nacion chilena"),
    ("D1", "D", "Libertario puro",
     r"estado minimo|libertad individual|libertario|partido nacional libertario|"
     r"\bpnl\b|anarcocapitalismo"),
]


def load_kast_speeches() -> pd.DataFrame:
    df = pd.read_csv(RAW_DISC)
    df["fecha"] = df["fecha"].map(parse_fecha_es)
    df["titulo"] = df["titulo"].fillna("").astype(str)
    df["discurso"] = df["discurso"].fillna("").astype(str)
    df = df[df["fecha"].notna()]
    df = df[df["fecha"] >= GOBIERNO_KAST]
    df = df[~df["titulo"].str.contains("Boric", case=False, regex=False)]
    df = df[df["discurso"].str.len() > 80]
    df["txt"] = df["discurso"].map(fold)
    df["mes"] = pd.to_datetime(df["fecha"]).dt.to_period("M").dt.to_timestamp()
    return df.reset_index(drop=True)


def top_palabras(kast: pd.DataFrame, n: int = 30) -> pd.DataFrame:
    tok_re = re.compile(r"[a-z]{4,}")
    cnt: dict[str, int] = {}
    for t in kast["txt"]:
        for w in tok_re.findall(t):
            if w in STOP or w.startswith(STOP_PREFIX):
                continue
            cnt[w] = cnt.get(w, 0) + 1
    rows = sorted(cnt.items(), key=lambda x: -x[1])[:n]
    return pd.DataFrame(rows, columns=["palabra", "n"])


def alusiones(kast: pd.DataFrame) -> pd.DataFrame:
    rows = []
    n = len(kast)
    for etiqueta, pat in ALUSIONES:
        hit = kast["txt"].str.contains(pat, regex=True)
        rows.append({"etiqueta": etiqueta, "n": int(hit.sum()), "pct": 100 * hit.mean(),
                     "n_discursos": n})
    return pd.DataFrame(rows)


def repertorios(kast: pd.DataFrame) -> pd.DataFrame:
    rows = []
    n = len(kast)
    for codigo, familia, etiqueta, pat in DICC:
        hit = kast["txt"].str.contains(pat, regex=True)
        rows.append({
            "codigo": codigo, "familia": familia, "etiqueta": etiqueta,
            "n": int(hit.sum()), "pct": 100 * hit.mean(), "n_discursos": n,
        })
    return pd.DataFrame(rows)


def mes_kast_kaiser(kast: pd.DataFrame) -> pd.DataFrame:
    g = kast.groupby("mes", as_index=False).agg(
        n_disc=("txt", "size"),
        n_kast=("txt", lambda s: int(s.str.contains(r"\bkast\b").sum())),
        n_kaiser=("txt", lambda s: int(s.str.contains(r"\bkaiser\b").sum())),
    )
    g["pct_kast"] = 100 * g["n_kast"] / g["n_disc"]
    g["pct_kaiser"] = 100 * g["n_kaiser"] / g["n_disc"]
    return g


def prensa_kast_kaiser() -> pd.DataFrame:
    pr = pd.read_parquet(PRENSA_UNIDA, columns=["year", "titulo"])
    pr = pr[pr["year"].between(2015, 2026)]
    tit = pr["titulo"].fillna("").map(fold)
    pr = pr.assign(
        dice_kast=tit.str.contains(r"\bkast\b", regex=True),
        dice_kaiser=tit.str.contains(r"\bkaiser\b", regex=True),
    )
    anio = (
        pr.groupby("year", as_index=False)
        .agg(n=("year", "size"), n_kast=("dice_kast", "sum"), n_kaiser=("dice_kaiser", "sum"))
    )
    anio["pct_kast"] = 100 * anio["n_kast"] / anio["n"]
    anio["pct_kaiser"] = 100 * anio["n_kaiser"] / anio["n"]
    return anio


def kaiser_context(kast: pd.DataFrame) -> str:
    hit = kast[kast["txt"].str.contains(r"\bkaiser\b")]
    if hit.empty:
        return ""
    t = hit.iloc[0]["txt"]
    i = t.find("kaiser")
    return t[max(0, i - 120): i + 140]


def main() -> int:
    OUT_CANON.mkdir(parents=True, exist_ok=True)
    OUT_PRENSA.mkdir(parents=True, exist_ok=True)

    kast = load_kast_speeches()
    print(f"Discursos Kast: {len(kast)}  {kast['fecha'].min()} → {kast['fecha'].max()}")

    top = top_palabras(kast)
    al = alusiones(kast)
    rep = repertorios(kast)
    mes = mes_kast_kaiser(kast)

    al.to_csv(OUT_CANON / "discursos_kast_alusiones.csv", index=False)
    mes.to_csv(OUT_CANON / "discursos_kast_kaiser_mes.csv", index=False)
    rep.to_csv(OUT_CANON / "discursos_kast_repertorios.csv", index=False)
    top.to_csv(OUT_CANON / "discursos_kast_top_palabras.csv", index=False)

    ctx = kaiser_context(kast)
    print("\nAlusiones (% discursos):")
    print(al.sort_values("pct", ascending=False).to_string(index=False, formatters={"pct": "{:.1f}".format}))
    print("\nRepertorios:")
    print(rep.sort_values("pct", ascending=False).to_string(index=False, formatters={"pct": "{:.1f}".format}))
    print("\nKast vs Kaiser por mes (cuerpo):")
    print(mes.to_string(index=False, formatters={"pct_kast": "{:.1f}".format, "pct_kaiser": "{:.1f}".format}))
    if ctx:
        print("\nÚnica mención a Kaiser:", ctx.replace("\n", " "))

    if PRENSA_UNIDA.exists():
        print("\nPrensa unida: Kast vs Kaiser por año (título)…")
        anio = prensa_kast_kaiser()
        anio.to_csv(OUT_PRENSA / "prensa_kast_kaiser_anio.csv", index=False)
        print(anio.to_string(index=False, formatters={
            "pct_kast": "{:.2f}".format, "pct_kaiser": "{:.2f}".format,
        }))
    else:
        print("No hay prensa_unida.parquet — corre python3 data/scripts/merge_fondecyt_neo.py")

    print("CSV listos → ggplot2: Rscript analysis/discursos_presidenciales/03_kast_alusiones.R")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
